eluthia/zipapp_script.py

```diff
@@ -0,0 +1,157 @@
+    """
+    Verifies that packages are installed and working correctly.
+    Input:
+        dict new_packages - new packages to be installed and their versions. Format {package_name: version}
+        int preinst_file_count - Number of files that were in badtrack's history folder before installation.
+        int timeout - how long to give badtrack to make it's first history file before rolling back.
+        int check_frequency - how often to check the history folder for new files.
+    Output:
+        bool - Whether the installation was successful.
+        string error_message - If installation was not successful, contains the reason why installation was detected to be unsuccessful.
+    """
+    def verify_badtrack(preinst_file_count, timeout, check_frequency):
+        while time.time() < timeout: # Verify that badtrack is making files
+            time.sleep(check_frequency)
+            if len(os.listdir(HISTORY_PATH)) > preinst_file_count:
+                return True
+
+        return False
+
+    timeout = time.time() + timeout
+
+    missing_packages = [package for package in new_packages.keys() if get_version(package) == None] # packages in new_packages that aren't installed
+    if missing_packages:
+            return False, f'Packages were not installed: {missing_packages}'
+
+    if not verify_badtrack(preinst_file_count, timeout, check_frequency):
+        return False, 'No files were generated by badtrack within 1 minute'
+
+    return True, None
+
+
+def rollback(new_packages, old_packages):
+    """
+    Rolls back the installation of packages.
+    Input:
+        dict new_packages - new packages to be installed and their versions. Format {package_name: version}
+        dict old_packages - old packages to be replaced and their versions. Format {package_name: version}
+    """
+    subprocess.run(['dpkg', '-r'] + new_packages.keys()) # remove new packages
+    reinstall_command = ['dpkg', '-E', '-i'] + [f'{OLD_DEB_PATH}/{package}.deb' for package in old_packages.keys()] # Necessary because of dependencies
+    subprocess.run(reinstall_command)
+    return
+
+
+def main():
+    os.makedirs(NEW_DEB_PATH, exist_ok=True)
+    os.makedirs(OLD_DEB_PATH, exist_ok=True)
+
+    # Extract archive, get packages and their versions
+    new_packages, old_packages = unpack_zip(os.path.dirname(__file__)) # the variables are dictionaries {name: version}
+
+
+    # Repack old packages
+    subprocess.run(['apt-get', 'install', 'dpkg-repack'])
+    for package_name, old_version in old_packages.items():
+        repack(package_name, old_version)
+
+    # Get current amount of badtrack history files (To verify that they increase after installation.)
+    preinst_file_count = len(os.listdir(HISTORY_PATH))
+
+    # Install new packages
+    install_command = ['dpkg', '-E', '-i'] + [f'{NEW_DEB_PATH}/{package}.deb' for package in new_packages.keys()]
```

Comment:
Comment on purpose of '-E'  argument and use a long argument name version if any.